### DISCUSSION TITLE
CSLA 6: Using services within a PropertyRuleAsync

### DISCUSSION THREAD

**Discussion by peranborkett** [MarimerLLC/csla#2719] [Ideas]
CSLA 6 Blazor Wasm/HttpPortal 

I'm updating an older code base, is there a recommended approach to injecting services into a async property rule for CSLA 6.

Consider the following simplified example:

`
protected override void AddBusinessRules()
{
  BusinessRules.AddRule(new ProductCodeUniqueAsyncRule(ProductCodeProperty, ApplicationContext.CreateInstanceDI<IDataPortal<ProductCodeUniqueCommand>>));
}
`

First issue would be that CreateInstanceDI<T> can not create an instance if T is an interface (I have a workaround for this, but not to get distracted from the main point...)
Second issue is even if I can inject the service in the constructor, subsequent calls to the dataportal use a different container scope and when the RevalidatingInterceptor runs the rules, the service throws an exception due to some objects already being disposed. (Business rules are instantiated once and reused across scopes)

My thoughts are that we need to inject the scoped service each time the rule is run (PropertyRuleAsync.ExecuteAsync())

RuleContext could be modified to allow services to be created from the scoped service provider, something like below.

`
IRuleContext.GetRequiredService<IDataPortal<ProductCodeUniqueCommand>>()
`

RuleContext has a private scoped ApplicationContext so should be doable.

**Comment by rockfordlhotka** [comment]
The next preview will include a `GetDataPortal` method for business and auth rules. The method is on the context, and will enable you to get access to an `IDataPortal<T>` that can be used to invoke commands or do other operations. https://github.com/MarimerLLC/csla/issues/2682

**Reply by peranborkett** [reply]
Fantastic, thank you.
